File: agent/tools/sandbox.py
"""Sandboxed shell execution for the agent's own bash/write/edit/read
tools -- not for the deterministic checks/git operations (agent/tools/
checks.py, agent/tools/git.py), which stay on the host unchanged since
those are fixed commands our code runs, never LLM-chosen, and already have
no isolation concern (nothing an LLM decided goes into them).

Running bash/write/edit directly on the same host as the agent server,
against real repo checkouts, is a real risk: a context-injected or
badly-confused agent could otherwise read or touch anything outside its
intended repo -- another project's live secrets on the same host, this
project's own .env, other tooling config directories, etc. Docker
containers, bind-mounting only the target repo's sandbox checkout, close
that specific risk. Local
containers were chosen over a remote sandbox provider specifically to avoid
a new third-party account/credentials. Network access stays enabled
(package installs still need it) -- the isolation boundary here is
filesystem, not network, which is still a meaningful improvement over no
boundary at all, not a claim of complete sandboxing.

Same {"ok", "exit_code", "output"} / ShellTimeout contract as
agent/tools/shell.py's run_shell, so agent_tools.py's tool functions barely
change to use this instead.
"""

import asyncio
from agent import runtime_settings as _rs
import logging
import os
import uuid

from agent.tools.shell import ShellTimeout

logger = logging.getLogger("tektonix")

# Every `-v` source below is resolved by the DOCKER DAEMON, not by this
# process -- which is the same thing while the agent runs directly on the
# host, and stops being the same thing the moment it runs in a container with
# the host's Docker socket mounted (docs/roadmap-packaging.md, M1). There the
# sandbox containers are SIBLINGS created by the host daemon: a `-v` naming a
# path that exists only inside the agent container silently mounts an empty
# directory, and the task reads an empty repo with no error anywhere.
#
# So the mapping is explicit. AGENT_HOST_PATH_MAP is a colon-free,
# comma-separated list of `container_prefix=host_prefix` pairs; unset (the
# normal single-host install) means the identity mapping, which is exactly
# today's behaviour.
#
#   AGENT_HOST_PATH_MAP=/projects=C:\dev,/workspaces=D:\work
_HOST_PATH_MAP_ENV = "AGENT_HOST_PATH_MAP"


def _host_path_map() -> list[tuple[str, str]]:
    """Parsed prefix pairs, longest container prefix first so a nested mapping
    wins over the parent it sits under."""
    raw = os.environ.get(_HOST_PATH_MAP_ENV, "").strip()
    pairs: list[tuple[str, str]] = []
    for entry in raw.split(","):
        entry = entry.strip()
        if not entry or "=" not in entry:
            continue
        container, host = entry.split("=", 1)
        container, host = container.strip().rstrip("/"), host.strip()
        if container and host:
            pairs.append((container, host.rstrip("/\\")))
    pairs.sort(key=lambda p: len(p[0]), reverse=True)
    return pairs


def host_path(path: str) -> str:
    """The path the Docker daemon should be given for a bind mount.

    Identity unless AGENT_HOST_PATH_MAP says otherwise. Only whole path
    SEGMENTS match, so a prefix of "/projects" never rewrites "/projects-old".
    """
    if not path:
        return path
    for container, host in _host_path_map():
        if path == container:
            return host
        if path.startswith(container + "/"):
            rest = path[len(container) + 1:]
            sep = "\\" if ("\\" in host or ":" in host[:2]) else "/"
            return f"{host}{sep}{rest.replace('/', sep)}"
    return path


SANDBOX_IMAGE = "tektonix-sandbox:latest"  # built from docker/agent-sandbox/Dockerfile
SANDBOX_MEMORY_LIMIT = "2g"
SANDBOX_CPU_LIMIT = "2"
# audit M-10: cap process count to blunt a fork bomb, drop all Linux
# capabilities (git/npm/node need none), and forbid privilege escalation.
SANDBOX_PIDS_LIMIT = "512"


async def _kill_container(name: str) -> None:
    # `docker run --rm`'s own cleanup only fires on a normal exit -- killing
    # the `docker run` CLI process (the asyncio subprocess we actually hold
    # a handle to) does not reliably stop the container itself; without this,
    # a timed-out or cancelled command would leave the container (and
    # whatever it's running) orphaned, running detached on the host --
    # exactly the same class of bug run_shell's own _kill_process_group
    # exists to prevent, just one layer up (container instead of process
    # group). `docker kill` + `--rm` together tear it down completely.
    try:
        proc = await asyncio.create_subprocess_exec(
            "docker", "kill", name,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        await proc.wait()
    except Exception:  # noqa: BLE001 -- best-effort cleanup, never worth failing the caller over
        pass


def _mount_allow_roots(cwd: str) -> list[str]:
    """Host paths a bind mount for this workspace may point at.

    Both mount computations below read state OUT of the worktree -- a
    node_modules symlink, and the .git pointer file -- and the worktree is
    bind-mounted read-write for the agent. Left unchecked that lets the agent
    choose what the host mounts into its own container: `ln -s /home
    <ws>/node_modules` yields `-v /home:/home:ro`, and a rewritten .git
    pointer yields `-v /root:/root:ro`. That is read access to every other
    project's .env, this agent's own secrets and ~/.ssh -- exactly what the
    module docstring claims to prevent.

    So a target is only accepted if it resolves under the workspace itself or
    under the LIVE checkout of the project this workspace belongs to, both of
    which come from server-owned config rather than from the worktree.

    An unconfigured workspace (a test fixture, a one-off checkout) gets the
    workspace alone -- fail closed, since there is no trusted second path to
    compare against.
    """
    roots = [os.path.realpath(cwd)]
    try:
        from agent.config import PROJECTS
    except Exception:  # noqa: BLE001 -- config unavailable; workspace-only
        return roots
    real_cwd = os.path.realpath(cwd)
    for cfg in PROJECTS.values():
        if os.path.realpath(cfg.get("sandbox", "")) == real_cwd:
            live = cfg.get("live")
            if live:
                roots.append(os.path.realpath(live))
            break
    return roots


def _mount_target_allowed(target: str, allow_roots: list[str]) -> bool:
    real = os.path.realpath(target)
    return any(real == r or real.startswith(r + os.sep) for r in allow_roots)


def _node_modules_mounts(cwd: str) -> list[str]:
    """audit C-2: some project worktrees symlink their `node_modules` (and, in a
    monorepo, nested ones) to the LIVE checkout's node_modules OUTSIDE the
    worktree -- e.g. a Next.js project'
    `<worktree>/node_modules -> /home/a Next.js project/node_modules`. Mounting only
    the worktree leaves that symlink dangling inside the container, so `npm run
    lint` fails with "eslint: not found" even though it passes on the host. For
    each such symlink whose target lives outside `cwd`, bind-mount the target
    read-only at its own absolute path so the symlink resolves. Read-only: the
    check phase must never mutate the live install. Real (non-symlink)
    node_modules dirs are already covered by the /workspace mount and need
    nothing here.
    """
    mounts: list[str] = []
    seen: set[str] = set()
    real_cwd = os.path.realpath(cwd)
    allow_roots = _mount_allow_roots(cwd)
    for root, dirs, _files in os.walk(cwd):
        # don't descend into the store internals; one level of nesting is enough
        # for the monorepo app/package node_modules layout.
        depth = root[len(cwd):].count(os.sep)
        if depth >= 3:
            dirs[:] = []
        if "node_modules" in dirs:
            nm = os.path.join(root, "node_modules")
            if os.path.islink(nm):
                target = os.path.realpath(nm)
                # only mount if it points OUTSIDE the worktree and actually exists
                if target and target not in seen \
                        and os.path.isdir(target) \
                        and not target.startswith(real_cwd + os.sep):
                    # audit C2: the symlink is agent-writable, so its target is
                    # an agent-chosen mount unless it is checked against
                    # server-owned paths.
                    if not _mount_target_allowed(target, allow_roots):
                        logger.warning(
                            "refusing to mount %s for %s: outside this project's "
                            "workspace and live checkout", target, cwd)
                        continue
                    seen.add(target)
                    mounts += ["-v", f"{host_path(target)}:{target}:ro"]
            # never walk into a node_modules tree
            dirs.remove("node_modules")
    return mounts


async def run_shell_sandboxed(
    cmd: str,
    cwd: str,
    # None means "use the operator-set default"; an explicit value from the
    # call site still wins, so the long-running checks keep their own caps.
    timeout: int | None = None,
    extra_env: dict | None = None,
    network: str | None = None,
) -> dict:
    """Runs `cmd` inside an ephemeral Docker container with only `cwd`
    (the repo sandbox checkout) bind-mounted, read-write, at /workspace --
    no other host path is visible inside the container. No TTY, CI=true,
    hard timeout -- same non-interactive-by-construction discipline as
    run_shell (see that module's own docstring on why: a pty-based shell tool
    can hang forever on an unanswered interactive prompt).

    `network` maps to docker's --network (e.g. "none" for full network
    isolation -- used by the deterministic check runner, whose deps are already
    installed so it needs no egress; audit C-2/M-10). None keeps the default
    bridge, which is what the interactive `bash` tool uses.

    Returns {"ok": bool, "exit_code": int, "output": str} -- same shape as
    run_shell, so callers don't need to branch on which one they used.
    """
    container_name = f"lga-{uuid.uuid4().hex[:12]}"
    env_args = []
    for k, v in {"CI": "true", "DEBIAN_FRONTEND": "noninteractive", **(extra_env or {})}.items():
        env_args += ["-e", f"{k}={v}"]

    # The workspace is a git WORKTREE of the live repo (see the 2026-08-25
    # workspace migration), and a worktree's .git is not a directory but a one-
    # line file: `gitdir: <live>/.git/worktrees/<name>`. Mounting only the
    # worktree therefore broke every git command inside the container — status,
    # diff, log all died with "fatal: not a git repository" because the pointer
    # target didn't exist in the mount namespace. It went unnoticed because the
    # deterministic git operations (agent/tools/git.py) run on the HOST; only
    # LLM-issued `git ...` in the bash tool was affected.
    #
    # Fix: when cwd is a worktree, also mount the live repo's .git at the SAME
    # absolute path, read-only. Same path, because that's what the pointer file
    # names; read-only, because the sandbox runs unreviewed model commands and
    # must not be able to rewrite the live repository's history — commits happen
    # on the host through verify_and_ship, never in here. Read operations work;
    # an in-sandbox `git commit` fails on the read-only mount, which is correct.
    git_mount_args: list[str] = []
    dotgit = os.path.join(cwd, ".git")
    if os.path.isfile(dotgit):
        try:
            with open(dotgit) as fh:
                gitdir = fh.read().strip().removeprefix("gitdir:").strip()
            # <live>/.git/worktrees/<name>  ->  <live>/.git
            marker = f"{os.sep}worktrees{os.sep}"
            if marker in gitdir:
                main_git = gitdir[: gitdir.index(marker)]
                # audit C2: this path comes from a file inside the worktree,
                # which the agent can rewrite -- `gitdir: /root/x` would mount
                # /root read-only into the container. Same allow-root check the
                # node_modules mounts get.
                if os.path.isdir(main_git) and _mount_target_allowed(
                        main_git, _mount_allow_roots(cwd)):
                    git_mount_args = ["-v", f"{host_path(main_git)}:{main_git}:ro"]
                elif os.path.isdir(main_git):
                    logger.warning(
                        "refusing to mount git dir %s for %s: outside this "
                        "project's workspace and live checkout", main_git, cwd)
        except OSError:
            pass  # unreadable pointer file: run without git rather than not at all

    nm_mount_args = _node_modules_mounts(cwd)
    net_args = ["--network", network] if network else []

    docker_args = [
        "docker", "run", "--rm", "--name", container_name,
        "-v", f"{host_path(cwd)}:/workspace",
        *git_mount_args,
        *nm_mount_args,
        *net_args,
        "-w", "/workspace",
        "--memory", SANDBOX_MEMORY_LIMIT,
        "--cpus", SANDBOX_CPU_LIMIT,
        # audit M-10 hardening. --user is deliberately NOT set: the bind-mounted
        # worktree is root-owned on the host (pm2 runs as root), so a non-root
        # container user could not write to it. Network stays on the default
        # bridge (a dedicated network blocking link-local/RFC1918 and binding
        # llm-router to 127.0.0.1 is tracked as residual infra) -- but
        # MODEL_ROUTER_KEY is not passed in, so the reachable router surface is
        # authenticated, not open.
        "--pids-limit", SANDBOX_PIDS_LIMIT,
        "--cap-drop", "ALL",
        "--security-opt", "no-new-privileges",
        *env_args,
        SANDBOX_IMAGE,
        "bash", "-c", cmd,
    ]
    proc = await asyncio.create_subprocess_exec(
        *docker_args,
        stdin=asyncio.subprocess.DEVNULL,  # no stdin at all -- matches run_shell's own reasoning
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.STDOUT,
    )
    try:
        _timeout = timeout if timeout is not None else _rs.as_int("sandbox_command_timeout_s")
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=_timeout)
    except TimeoutError:
        await _kill_container(container_name)
        await proc.wait()
        # _timeout, not timeout: the latter is None when the caller took the
        # operator default, which would report "timed out after None seconds".
        raise ShellTimeout(cmd, _timeout)
    except asyncio.CancelledError:
        # Same reasoning as run_shell's own CancelledError handler: the
        # operator's Stop button cancels the asyncio task driving the whole
        # run, and without explicit cleanup here the container (and
        # anything running inside it) would keep running orphaned.
        await _kill_container(container_name)
        await proc.wait()
        raise

    output = stdout.decode("utf-8", errors="replace")
    return {"ok": proc.returncode == 0, "exit_code": proc.returncode, "output": output[-20_000:]}

File: services/commit-reviewer/sandbox.js
// Running agent-authored code somewhere it cannot reach this machine.
//
// THE PROBLEM. reviewProject runs each project's configured checks -- `npm
// test` and friends -- against the agent's worktree, with execFile, on the
// host, as root. The agent's own shell is sandboxed, but its write/edit tools
// put files in that worktree, so a test file it wrote is arbitrary code
// executing as root outside any container. sealedEnv() stops secrets leaking
// INTO those commands; it does nothing about what they can do once running.
// That is a complete path from "model in a container" to "root on the box",
// and it is what this module closes.
//
// WHY THIS IS NOT A LEAP. The commands are not new to a container. The agent
// already runs the same cfg.checks list, in the same worktree, through
// agent/tools/sandbox.py's run_shell_sandboxed with --network none, and its
// run_checks tool describes them as "the exact same commands the outer
// verification gate will run". Every project that has completed a task has
// already proven its checks work in this image. The flags below deliberately
// mirror that call site rather than inventing a second hardening policy --
// two policies drift, and the weaker one is the one that matters.
//
// WHY THE BUNDLE IS EXEMPT. In the compose bundle this service runs INSIDE a
// container that is not given /var/run/docker.sock (docker-compose.yml gives
// it to `agent` alone). It cannot start a container, and handing it the
// socket so that it could would give that container host-root equivalent --
// strictly worse than the containment it already has. So on the bundle the
// caller keeps running checks in-process, and only a host install, where the
// reviewer really is root on the real machine, routes through here.
//
// FAIL CLOSED. If Docker or the image is missing on a host install, a check
// does not quietly run on the host instead: "fall back when the sandbox
// fails" is the path anyone attacking this would engineer. It is not a new
// fragility either -- the agent's own bash already requires Docker, so a box
// without it is not running tasks to review.
'use strict';

const { execFile } = require('child_process');
const fs = require('fs');
const path = require('path');

// Mirrors agent/tools/sandbox.py. Kept as literals with the source named
// rather than read from Python, because a reviewer that cannot parse the
// agent's config should still run with the documented limits.
// One copy of the stack->image map, shared with scripts/verify_stack_checks.py
// and agent/provisioning.py. See docker/stack-images.json for why.
let STACKS = { default: 'tektonix-sandbox:latest', stacks: {} };
try {
    STACKS = JSON.parse(fs.readFileSync(
        path.join(__dirname, '..', '..', 'docker', 'stack-images.json'), 'utf8'));
} catch {
    // A missing map is not fatal: every project without a declared stack runs
    // in the default image, which is the behaviour that existed before stacks
    // were selectable at all.
}

const IMAGE = process.env.AGENT_SANDBOX_IMAGE || STACKS.default || 'tektonix-sandbox:latest';

/**
 * The image and extra env a check runs with.
 *
 * A check may name its own `stack` -- set at onboarding from what
 * provisioning detected. Per CHECK, not per project, because a Go backend
 * with a React frontend is an ordinary repository and its two checks belong
 * in two different images.
 *
 * An unknown stack falls back to the default rather than failing: the map can
 * gain entries after a project's config was written, and refusing to run a
 * check because its label is new is worse than running it where it probably
 * works.
 */
function imageFor(stack) {
    const entry = stack && STACKS.stacks ? STACKS.stacks[stack] : null;
    return {
        image: (entry && entry.image) || IMAGE,
        env: (entry && entry.env) || {},
        known: Boolean(entry),
    };
}
const MEMORY = '2g';
const CPUS = '2';
const PIDS = '512';

const IN_CONTAINER = process.env.TEKTONIX_BUNDLE === '1';

function execp(cmd, args, opts = {}) {
    return new Promise((resolve) => {
        execFile(cmd, args, { maxBuffer: 20 * 1024 * 1024, ...opts }, (err, stdout, stderr) => {
            resolve({ ok: !err, code: err ? (err.code ?? 1) : 0, out: (stdout || '') + (stderr || '') });
        });
    });
}

let _probe = null;

/**
 * Whether checks can be sandboxed here, cached for the process.
 *
 * Three answers, not two: `bundle` is "already contained, do not try", which
 * a caller must not confuse with `unavailable` -- one is correct operation
 * and the other is a host install that has to refuse.
 */
async function probe() {
    if (_probe) return _probe;
    if (IN_CONTAINER) {
        _probe = { mode: 'bundle', reason: 'this service runs in a container that is not given the docker socket' };
        return _probe;
    }
    const d = await execp('docker', ['version', '--format', '{{.Server.Version}}']);
    if (!d.ok) {
        _probe = { mode: 'unavailable', reason: `docker is not usable here: ${d.out.trim().slice(0, 200)}` };
        return _probe;
    }
    const img = await execp('docker', ['image', 'inspect', IMAGE, '--format', '{{.Id}}']);
    if (!img.ok) {
        _probe = { mode: 'unavailable', reason: `the sandbox image ${IMAGE} is not built on this host` };
        return _probe;
    }
    _probe = { mode: 'sandbox', reason: `${IMAGE} on docker ${d.out.trim()}` };
    return _probe;
}

function resetProbe() { _probe = null; }   // tests only

/**
 * The -v arguments a worktree needs for its checks to behave as they do on
 * the host.
 *
 * A plain `-v <worktree>:/workspace` is not enough, and getting this wrong is
 * how the change would have broken every project: the reviewer brings
 * dependencies in with `mount --bind` on the host, and a bind mount NESTED
 * inside a directory is not carried into a container by a plain bind of its
 * parent. node_modules would be an empty directory inside, and every `npm
 * test` would fail with "not found" while passing on the host.
 *
 * So each declared dependency and read-only mount is bound explicitly from
 * its source on the live checkout -- the same source the host bind uses --
 * read-only, because a check must never write into live's installed
 * dependencies or its data.
 *
 * The worktree's .git is a pointer FILE to <live>/.git/worktrees/<name>, so
 * git inside the container needs the live .git at the same absolute path, or
 * every `git` in a check dies with "not a git repository".
 */
function isInside(child, parent) {
    const rel = path.relative(path.resolve(parent), path.resolve(child));
    return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel));
}

/** Relative paths of symlinked node_modules directories, up to three deep. */
function nodeModulesLinks(root, depth = 0, rel = '') {
    if (depth > 3) return [];
    const out = [];
    let entries;
    try {
        entries = fs.readdirSync(path.join(root, rel), { withFileTypes: true });
    } catch {
        return out;
    }
    for (const e of entries) {
        const here = path.join(rel, e.name);
        if (e.name === 'node_modules') {
            if (e.isSymbolicLink()) out.push(here);
            continue;   // never descend into one
        }
        if (e.isDirectory() && !e.name.startsWith('.')) {
            out.push(...nodeModulesLinks(root, depth + 1, here));
        }
    }
    return out;
}

function mountArgs(cfg, worktreePath) {
    const args = ['-v', `${worktreePath}:/workspace`];
    const seen = new Set();

    for (const rel of [...(cfg.dependencyDirs || []), ...(cfg.readOnlyMounts || [])]) {
        const src = path.join(cfg.live, rel);
        if (seen.has(rel) || !fs.existsSync(src)) continue;
        seen.add(rel);
        args.push('-v', `${src}:${path.posix.join('/workspace', rel)}:ro`);
    }

    // The other half of the same problem, and the one that actually bit:
    // some worktrees SYMLINK node_modules to the live checkout's copy rather
    // than holding their own. A symlink is carried into the container
    // faithfully and then dangles, so `npx eslint` dies with "not found"
    // inside while passing on the host. Mount each such target read-only at
    // its own absolute path so the link resolves -- the same fix, and the
    // same reasoning, as _node_modules_mounts in agent/tools/sandbox.py.
    //
    // Read-only because a check must never write into live's installed
    // dependencies. Bounded to three levels: deep enough for a monorepo's
    // per-package node_modules, shallow enough not to walk a whole tree.
    for (const rel of nodeModulesLinks(worktreePath)) {
        // A DANGLING link throws here rather than resolving, and an agent that
        // deleted a directory its symlink pointed at is an ordinary state for
        // a worktree to be in. Unguarded, that ENOENT came out of mountArgs
        // and failed the entire review rather than skipping one mount.
        let target;
        try {
            target = fs.realpathSync(path.join(worktreePath, rel));
        } catch {
            continue;
        }
        // Agent-writable worktree, so the link target is agent-controlled:
        // only the project's own live checkout is an acceptable destination.
        if (!isInside(target, cfg.live) || seen.has(target)) continue;
        seen.add(target);
        args.push('-v', `${target}:${target}:ro`);
    }

    // ...and the layout the loop above cannot see. When a project's package
    // directories carry named package.json files, setupWorktree builds a REAL
    // node_modules/ directory and symlinks each ENTRY inside it into live's
    // copy (so a workspace-internal package can be redirected to the
    // worktree's own build). node_modules itself is then not a link, the loop
    // above finds nothing, and every entry dangles inside the container.
    //
    // Found 2026-09-22 on a project of three standalone apps: once the review
    // service finally knew which directories needed node_modules, the checks
    // STILL failed with `oxlint: not found`, because frontend/node_modules/.bin
    // pointed at /home/<project>/frontend/node_modules/.bin -- a path the
    // container had never been given. Mounting each configured directory's
    // live node_modules read-only at its own absolute path makes every such
    // entry resolve, and it is the same mount the symlink layout already gets.
    for (const rel of cfg.nodeModulesDirs || []) {
        const liveNm = path.join(cfg.live, rel, 'node_modules');
        let target;
        try {
            target = fs.realpathSync(liveNm);
        } catch {
            continue;                       // not installed in live: nothing to mount
        }
        if (!isInside(target, cfg.live) || seen.has(target)) continue;
        seen.add(target);
        args.push('-v', `${target}:${target}:ro`);
    }

    const dotgit = path.join(worktreePath, '.git');
    try {
        if (fs.statSync(dotgit).isFile()) {
            const gitdir = fs.readFileSync(dotgit, 'utf8').replace(/^gitdir:/, '').trim();
            const marker = `${path.sep}worktrees${path.sep}`;
            if (gitdir.includes(marker)) {
                const mainGit = gitdir.slice(0, gitdir.indexOf(marker));
                // The pointer file lives inside a directory the agent can
                // write, so the path in it is agent-controlled: `gitdir:
                // /root/x` would otherwise mount /root into the container.
                // Only the live checkout's own .git is accepted.
                const expected = path.join(cfg.live, '.git');
                if (path.resolve(mainGit) === path.resolve(expected) && fs.existsSync(mainGit)) {
                    args.push('-v', `${mainGit}:${mainGit}:ro`);
                }
            }
        }
    } catch {
        // An unreadable pointer means git will not work inside; a check that
        // needs git fails loudly there rather than silently running on the host.
    }
    return args;
}

/**
 * Run one check command inside the sandbox.
 *
 * argv, never a shell: the reviewer's own runSealed uses execFile for the
 * same reason, and passing `cmd args...` to `bash -c` here would ADD an
 * injection surface the host path does not have.
 *
 * `--network none` by default, matching the agent's own check runner:
 * dependencies are installed by the time checks run, and nothing here may
 * install (see reviewProject's note on why an install would execute
 * untrusted code).
 *
 * A check may opt out with `network: "bridge"` in projects.json, and one
 * real check needs it -- `pnpm audit` queries an advisory database and fails
 * with no egress. That opt-in is deliberately in the project CONFIG, which
 * the agent cannot write: it lives outside the worktree, so a model cannot
 * grant its own code network access by editing a file. What the operator IS
 * agreeing to, per check, is that this one command runs agent-authored code
 * WITH egress -- which is why it is per check and not a global switch.
 */
/**
 * The exact argv handed to docker. Pure, and exported, so the hardening can
 * be ASSERTED rather than grepped for: --cap-drop ALL and --network none are
 * the difference between containment and a container, and a source grep
 * passes for a flag that has been moved into a branch that never runs.
 */
function dockerArgs(cfg, worktreePath, relDir, cmd, args, extraEnv, network, stack) {
    const target = imageFor(stack);
    const envArgs = [];
    for (const [k, v] of Object.entries({
        CI: 'true', DEBIAN_FRONTEND: 'noninteractive', LANG: 'C.UTF-8',
        // The toolchain's own needs first, so a check may still override
        // them: with no network and a read-only HOME, Go wants a GOCACHE it
        // can write and Maven a local repository, or nothing runs at all.
        ...target.env, ...(extraEnv || {}),
    })) {
        envArgs.push('-e', `${k}=${v}`);
    }

    const docker = [
        'run', '--rm',
        ...mountArgs(cfg, worktreePath),
        '--network', network === 'bridge' ? 'bridge' : 'none',
        '-w', path.posix.join('/workspace', relDir || '.'),
        '--memory', MEMORY,
        '--cpus', CPUS,
        '--pids-limit', PIDS,
        '--cap-drop', 'ALL',
        '--security-opt', 'no-new-privileges',
        // --entrypoint, not `IMAGE cmd args`. This image is built on the
        // Node base, which ships ENTRYPOINT ["docker-entrypoint.sh"]; that
        // script hands anything it does not recognise to `node`, so a check
        // whose tool is missing came back as a Node module stack trace
        // instead of "not found". Setting the entrypoint bypasses it and
        // gives docker's own, unambiguous:
        //   exec: "go": executable file not found in $PATH
        // which is what lets a missing toolchain be reported as a setup
        // problem rather than as a failing check.
        '--entrypoint', cmd,
        ...envArgs,
        target.image,
        ...(args || []),
    ];

    return { docker, image: target.image };
}


async function runSandboxed(cfg, worktreePath, relDir, cmd, args, timeoutMs, extraEnv, network, stack) {
    const { docker, image } = dockerArgs(cfg, worktreePath, relDir, cmd, args, extraEnv, network, stack);
    const r = await execp('docker', docker, { timeout: timeoutMs || 300_000 });
    const missing = missingTool(r.out);
    if (missing) {
        // A toolchain the image does not carry is a SETUP problem, not a
        // failing check, and the difference decides who fixes it. Left as a
        // raw docker error it reads as "the agent broke the build", the
        // agent is asked to debug an environment it cannot see, and it tries
        // -- which is how a correct commit gets rejected round after round
        // (the same reasoning as MISSING_TOOL_RE in reviewer.js).
        return {
            ok: false,
            code: r.code,
            missingTool: missing,
            output: `SETUP: this check needs \`${missing}\`, which the sandbox image `
                  + `(${image}) does not have. The image carries Node and Python; a project on `
                  + `another toolchain needs it added to docker/agent-sandbox/Dockerfile, or this `
                  + `check removed from the project's config. Nothing about the code under review `
                  + `is known either way -- the check did not run.`,
        };
    }
    return { ok: r.ok, output: r.out, code: r.code };
}

// Docker's own words when the image has no such binary. Deliberately narrow:
// it must not match a program that merely PRINTS something similar, because
// mislabelling a real failure as a setup problem hides a genuine break.
const NOT_IN_IMAGE_RE = /exec: "([^"]+)": executable file not found in \$PATH/;

function missingTool(output) {
    const m = NOT_IN_IMAGE_RE.exec(output || '');
    return m ? m[1] : null;
}

module.exports = { probe, resetProbe, runSandboxed, dockerArgs, mountArgs, nodeModulesLinks, isInside,
                   missingTool, imageFor, STACKS, IMAGE, IN_CONTAINER };
